### PR TITLE
Fixing various virtual memory issues

### PR DESCRIPTION
**Pull request by supercomputer7**
This PR fixes:

- `E1000` Network card driver, so instead of doing identity-mapping, the driver will allocate a sized `Region` mapped to the MMIO physical address.
- The DMI decoder & the ACPI static parser are not using their virtual-memory mapping methods anymore, so they use a general method from `MemoryManager` instead.
- The PCI initialization code doesn't rely on the DMI decoder anymore. It was my mistake, though Bochs does inform that PCI is supported via SMBIOS, we still don't want to rely on the SMBIOS info for critical measurements.

However there are still some issues I can't find solutions for them.
I'm still not 100% sure this patch will work on all machines or emulators.
My QEMU 4.2.0 seems to work fine, but when I tested it with QEMU 4.1.0 it doesn't work.
I'll need to see what's going on with QEMU 4.1.0 before this PR gets merged.

### CODE REVIEW

**Comment by supercomputer7** [comment]
I noticed that it's not about QEMU 4.1.0 that doesn't work but some bug I can't diagnose about the Q35 machine type. i440fx machines work without a problem, and if there's a problem with the ACPI code in modern machines we can bypass with `noacpi` kernel boot parameter.
So technically speaking, this patch doesn't create a regression, if we can still use the i440fx machine type without any problem.

File: Kernel/Process.cpp [hunk 1]
```diff
@@ -64,6 +64,7 @@ static Lock* s_hostname_lock;
 static VirtualAddress s_info_page_address_for_userspace;
 static VirtualAddress s_info_page_address_for_kernel;
 VirtualAddress g_return_to_ring3_from_signal_trampoline;
+Region* g_return_to_ring3_from_signal_trampoline_region;
```
Comment (awesomekling) [inline Kernel/Process.cpp]: There's no need to keep this `Region` in a global variable, as it has no outside clients.

File: Kernel/VM/Region.cpp [hunk 2]
```diff
@@ -228,6 +231,7 @@ void Region::map_individual_page_impl(size_t page_index)
         pte.set_physical_page_base(0);
         pte.set_present(false);
     } else {
+        pte.set_cache_disabled(!m_cacheable);
```
Comment (awesomekling) [inline Kernel/VM/Region.cpp]: This change means that basically all memory now becomes uncacheable.
It's good that you're adding the ability to control this flag, but the `cacheable` argument to the `Region` construction helpers should really be `true` by default. If you declare those functions with `cacheable = true`, you also don't have to update the callers, making the patch a bit smaller :)

File: Kernel/Arch/i386/CPU.h [hunk 3]
```diff
@@ -421,7 +422,17 @@ struct [[gnu::aligned(16)]] FPUState
...
+
+inline constexpr u32 offset_in_page(u32 address)
+{
+    return address & (~PAGE_MASK);
+}
+
+inline VirtualAddress create_vm_address(VirtualAddress v_base, PhysicalAddress p_offset)
+{
+    return VirtualAddress(page_base_of(v_base.get()) + offset_in_page(p_offset.get()));
 }
```
Comment (awesomekling) [inline Kernel/Arch/i386/CPU.h]: This API mixes `VirtualAddress` and `PhysicalAddress`, when the purpose of those classes is to *not* mix virtual and physical address. :)

Use of this function can be replaced by existing `VirtualAddress` functionality:
```cpp
auto addr = v_base.page_base().offset(some_offset);
```

File: Kernel/ACPI/ACPIStaticParser.cpp [hunk 4]
```diff
@@ -117,39 +100,91 @@ void ACPIStaticParser::do_acpi_shutdown()
...
 #ifdef ACPI_DEBUG
-    dbgprintf("ACPI: Checking Main SDT Length to choose the correct mapping size");
+    dbgprintf("ACPI: Checking SDT Length\n");
 #endif
-    mmap_region(*checkup_region, PhysicalAddress((u32)m_main_system_description_table & PAGE_MASK));
+    auto region = MM.allocate_kernel_region(PhysicalAddress((u32)&p_header & PAGE_MASK), (PAGE_SIZE * 2), "ACPI get_table_size()", Region::Access::Read);
+    volatile auto* sdt = (ACPI_RAW::SDTHeader*)(create_vm_address(region->vaddr(), PhysicalAddress((u32)&p_header))).as_ptr();
+    // NOTE & FIXME: Apparently we need a loop to make sure we don't read some random garbage...
+    while (strncmp(const_cast<const char*>(sdt->sig), sig, strlen(sig)))
+        ;
```
Comment (awesomekling) [inline Kernel/ACPI/ACPIStaticParser.cpp]: What's going on here? This doesn't look sane.

File: Kernel/ACPI/ACPIStaticParser.cpp [hunk 5]
```diff
@@ -51,45 +52,27 @@ ACPI_RAW::SDTHeader* ACPIStaticParser::find_table(const char* sig)
...
 
-    auto checkup_region = MM.allocate_kernel_region((PAGE_SIZE * 2), "ACPI Static Parser", Region::Access::Read);
+    auto checkup_region = MM.allocate_kernel_region(PhysicalAddress(page_base_of((u32)(fadt_ptr))), (PAGE_SIZE * 2), "ACPI Static Parser", Region::Access::Read);
 #ifdef ACPI_DEBUG
     dbgprintf("ACPI: Checking FADT Length to choose the correct mapping size\n");
 #endif
-    mmap_region(*checkup_region, PhysicalAddress((u32)(&physical_fadt) & PAGE_MASK));
 
-    ACPI_RAW::SDTHeader* sdt = (ACPI_RAW::SDTHeader*)(checkup_region->vaddr().get() + ((u32)(&physical_fadt) & (~PAGE_MASK)));
+    ACPI_RAW::SDTHeader* sdt = (ACPI_RAW::SDTHeader*)(create_vm_address(checkup_region->vaddr(), PhysicalAddress((u32)(fadt_ptr)))).as_ptr();
```
Comment (awesomekling) [inline Kernel/ACPI/ACPIStaticParser.cpp]: Here's a much easier and shorter way to do this:

```suggestion
    auto* sdt = (ACPI_RAW::SDTHeader*)(checkup_region->vaddr().page_base().offset(fadt_ptr)).as_ptr();
```

File: Kernel/ACPI/ACPIStaticParser.cpp [hunk 6]
```diff
@@ -51,45 +52,27 @@ ACPI_RAW::SDTHeader* ACPIStaticParser::find_table(const char* sig)
...
+    dbgprintf("ACPI: FADT @ V 0x%x, P 0x%x\n", sdt, fadt_ptr);
 #endif
     u32 length = sdt->length;
     kprintf("ACPI: Fixed ACPI data, Revision %u\n", sdt->revision);
 
-    auto fadt_region = MM.allocate_kernel_region(PAGE_ROUND_UP(length) + PAGE_SIZE, "ACPI Static Parser", Region::Access::Read);
-    mmap_region(*fadt_region, PhysicalAddress((u32)(&physical_fadt) & PAGE_MASK));
-    m_fadt = OwnPtr<ACPI::FixedACPIData>(new ACPI::FixedACPIData(*((ACPI_RAW::FADT*)(fadt_region->vaddr().get() + ((u32)(&physical_fadt) & (~PAGE_MASK))))));
+    auto fadt_region = MM.allocate_kernel_region(PhysicalAddress(page_base_of((u32)(fadt_ptr))), PAGE_ROUND_UP(length) + PAGE_SIZE, "ACPI Static Parser", Region::Access::Read);
+    m_fadt = OwnPtr<ACPI::FixedACPIData>(new ACPI::FixedACPIData(*((ACPI_RAW::FADT*)create_vm_address(fadt_region->vaddr(), PhysicalAddress((u32)(fadt_ptr))).as_ptr())));
```
Comment (awesomekling) [inline Kernel/ACPI/ACPIStaticParser.cpp]:
```suggestion
    m_fadt = make<ACPI::FixedACPIData>(*(ACPI_RAW::FADT*)fadt_region->vaddr().page_base().offset((u32)fadt_ptr).as_ptr());
```

File: Kernel/ACPI/ACPIStaticParser.cpp [hunk 7]
```diff
@@ -117,39 +100,91 @@ void ACPIStaticParser::do_acpi_shutdown()
...
+{
+    InterruptDisabler disabler;
+#ifdef ACPI_DEBUG
+    dbgprintf("ACPI: Checking SDT Revision\n");
+#endif
+    auto region = MM.allocate_kernel_region(PhysicalAddress((u32)&p_header & PAGE_MASK), (PAGE_SIZE * 2), "ACPI get_table_revision()", Region::Access::Read);
+    volatile auto* sdt = (ACPI_RAW::SDTHeader*)(create_vm_address(region->vaddr(), PhysicalAddress((u32)&p_header))).as_ptr();
+    // NOTE & FIXME: Apparently we need a loop to make sure we don't read some random garbage...
+    while (strncmp(const_cast<const char*>(sdt->sig), sig, strlen(sig)))
+        ;
```
Comment (awesomekling) [inline Kernel/ACPI/ACPIStaticParser.cpp]: Same question here.

File: Kernel/ACPI/Definitions.h [hunk 8]
```diff
@@ -109,28 +116,44 @@ struct FADT {
...
 
-struct DSDT : public SDTHeader {
+struct [[gnu::packed]] AMLTable
+{
+    SDTHeader h;
+    char aml_code[];
 };
 
-struct PCI_MMIO_Descriptor {
+struct [[gnu::packed]] PCI_MMIO_Descriptor
```
Comment (awesomekling) [inline Kernel/ACPI/Definitions.h]: It's okay to change from `__attribute__((__packed__))` to `[[gnu::packed]]` but please do that sort of thing in a separate commit, since it makes the actual changes much harder to read.

File: Kernel/ACPI/ACPIStaticParser.cpp (hunk 5 above)
Comment (supercomputer7) [reply Kernel/ACPI/ACPIStaticParser.cpp]: `auto* sdt = (ACPI_RAW::SDTHeader*)(checkup_region->vaddr().page_base().offset(fadt_ptr)).as_ptr();
`
is wrong, since it won't trim the upper bits.

Comment (awesomekling) [reply Kernel/ACPI/ACPIStaticParser.cpp]: Ah yes, you are right! That's my mistake, we'll need to use `offset_in_page` as well :)

Comment (supercomputer7) [reply Kernel/ACPI/ACPIStaticParser.cpp]: OK, that was fixed, thanks :)

File: Kernel/ACPI/ACPIStaticParser.cpp (hunk 7 above)
Comment (supercomputer7) [reply Kernel/ACPI/ACPIStaticParser.cpp]: We discussed it in the IRC channel. Apparently, we didn't flush the paging tables though I assumed we did (that was the problem for days that I couldn't figure what's going on!).
Fixed.

File: Kernel/ACPI/ACPIStaticParser.cpp (hunk 4 above)
Comment (supercomputer7) [reply Kernel/ACPI/ACPIStaticParser.cpp]: We discussed it in the IRC channel. Apparently, we didn't flush the paging tables though I assumed we did (that was the problem for days that I couldn't figure what's going on!).
Fixed.

File: Kernel/ACPI/Definitions.h (hunk 8 above)
Comment (supercomputer7) [reply Kernel/ACPI/Definitions.h]: Sure, will do so :)

File: Kernel/Arch/i386/CPU.h (hunk 3 above)
Comment (supercomputer7) [reply Kernel/Arch/i386/CPU.h]: Yeah, your argument of "offsetting an Address with another is not making sense" does make sense :)
Fixed.

File: Kernel/Process.cpp (hunk 1 above)
Comment (supercomputer7) [reply Kernel/Process.cpp]: Fixed.

File: Kernel/VM/Region.cpp (hunk 2 above)
Comment (supercomputer7) [reply Kernel/VM/Region.cpp]: that's actually a good idea :)

File: Kernel/ACPI/ACPIStaticParser.cpp (hunk 6 above)
Comment (supercomputer7) [reply Kernel/ACPI/ACPIStaticParser.cpp]: ```fadt_region->vaddr().page_base().offset((u32)fadt_ptr).as_ptr()```
is wrong, since it won't trim the upper bits as I mentioned above.
But, using make<> is a nice suggestion :)

File: Kernel/VM/Region.cpp [hunk 9]
```diff
@@ -65,7 +68,7 @@ NonnullOwnPtr<Region> Region::clone()
             vaddr().get());
 #endif
         // Create a new region backed by the same VMObject.
-        auto region = Region::create_user_accessible(m_range, m_vmobject, m_offset_in_vmobject, m_name, m_access);
+        auto region = Region::create_user_accessible(m_range, m_vmobject, m_offset_in_vmobject, m_name, m_access, false);
```
Comment (awesomekling) [inline Kernel/VM/Region.cpp]: We should not make this region uncacheable.

File: Kernel/VM/Region.cpp [hunk 10]
```diff
@@ -80,7 +83,7 @@ NonnullOwnPtr<Region> Region::clone()
     // Set up a COW region. The parent (this) region becomes COW as well!
     ensure_cow_map().fill(true);
     remap();
-    auto clone_region = Region::create_user_accessible(m_range, m_vmobject->clone(), m_offset_in_vmobject, m_name, m_access);
+    auto clone_region = Region::create_user_accessible(m_range, m_vmobject->clone(), m_offset_in_vmobject, m_name, m_access, false);
```
Comment (awesomekling) [inline Kernel/VM/Region.cpp]: We should not make this region uncacheable.

**Comment by supercomputer7** [comment]
@awesomekling 
OK, I just need to move the big file changes (those where I changed `__attribute__((__packed__))` to `[[gnu::packed]]`)... 
If you have any suggestion, please let me know about it :)

**Comment by supercomputer7** [comment]
@awesomekling Thanks for the review, I fixed the mistakes :)
I remember I did git-clang-format on the code and I think it's fine, but I might be wrong about that.
